File: lang/c/src/vtfppc/PPL.cpp
#include <vtfppc/PPL.h>

#include <vtfpp/PPL.h>

#include <sourceppc/Helpers.h>
#include <vtfppc/Convert.hpp>

using namespace sourceppc;
using namespace vtfpp;

SOURCEPP_API vtfpp_ppl_handle_t vtfpp_ppl_create(uint32_t modelChecksum) {
	return new PPL{modelChecksum};
}

SOURCEPP_API vtfpp_ppl_handle_t vtfpp_ppl_create_with_options(uint32_t modelChecksum, vtfpp_image_format_e format, uint32_t version) {
	return new PPL{modelChecksum, static_cast<ImageFormat>(format), version};
}

SOURCEPP_API vtfpp_ppl_handle_t vtfpp_ppl_open_from_mem(const unsigned char* buffer, size_t bufferLen) {
	SOURCEPP_EARLY_RETURN_VAL(buffer, nullptr);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, nullptr);

	return new PPL{{reinterpret_cast<const std::byte*>(buffer), bufferLen}};
}

SOURCEPP_API vtfpp_ppl_handle_t vtfpp_ppl_open_from_file(const char* pplPath) {
	SOURCEPP_EARLY_RETURN_VAL(pplPath, nullptr);

	return new PPL{pplPath};
}

SOURCEPP_API int vtfpp_ppl_is_valid(vtfpp_ppl_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);

	return Convert::ppl(handle)->operator bool();
}

SOURCEPP_API uint32_t vtfpp_ppl_get_version(vtfpp_ppl_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return Convert::ppl(handle)->getVersion();
}

SOURCEPP_API void vtfpp_ppl_set_version(vtfpp_ppl_handle_t handle, uint32_t version) {
	SOURCEPP_EARLY_RETURN(handle);

	return Convert::ppl(handle)->setVersion(version);
}

SOURCEPP_API uint32_t vtfpp_ppl_get_model_checksum(vtfpp_ppl_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, 0);

	return Convert::ppl(handle)->getModelChecksum();
}

SOURCEPP_API void vtfpp_ppl_set_model_checksum(vtfpp_ppl_handle_t handle, uint32_t modelChecksum) {
	SOURCEPP_EARLY_RETURN(handle);

	Convert::ppl(handle)->setModelChecksum(modelChecksum);
}

SOURCEPP_API vtfpp_image_format_e vtfpp_ppl_get_format(vtfpp_ppl_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, VTFPP_IMAGE_FORMAT_EMPTY);

	return static_cast<vtfpp_image_format_e>(Convert::ppl(handle)->getFormat());
}

SOURCEPP_API void vtfpp_ppl_set_format(vtfpp_ppl_handle_t handle, vtfpp_image_format_e format, float quality) {
	SOURCEPP_EARLY_RETURN(handle);

	Convert::ppl(handle)->setFormat(static_cast<ImageFormat>(format), quality);
}

SOURCEPP_API int vtfpp_ppl_has_image_for_lod(vtfpp_ppl_handle_t handle, uint32_t lod) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);

	return Convert::ppl(handle)->hasImageForLOD(lod);
}

SOURCEPP_API sourcepp_buffer_t vtfpp_ppl_get_image_lods(vtfpp_ppl_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(Convert::ppl(handle)->getImageLODs());
}

SOURCEPP_API sourcepp_buffer_t vtfpp_ppl_get_image_raw(vtfpp_ppl_handle_t handle, uint32_t* width, uint32_t* height, uint32_t lod) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(Convert::ppl(handle)->getImageLODs());
}

SOURCEPP_API sourcepp_buffer_t vtfpp_ppl_get_image_as(vtfpp_ppl_handle_t handle, uint32_t* width, uint32_t* height, vtfpp_image_format_e format, uint32_t lod) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, SOURCEPP_BUFFER_INVALID);

	auto image = Convert::ppl(handle)->getImageAs(static_cast<ImageFormat>(format), lod);
	if (!image) {
		return SOURCEPP_BUFFER_INVALID;
	}
	*width = image->width;
	*height = image->height;
	return convert::toBuffer(image->data);
}

SOURCEPP_API sourcepp_buffer_t vtfpp_ppl_get_image_as_rgb888(vtfpp_ppl_handle_t handle, uint32_t* width, uint32_t* height, uint32_t lod) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(width, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(height, SOURCEPP_BUFFER_INVALID);

	auto image = Convert::ppl(handle)->getImageAs(ImageFormat::RGB888, lod);
	if (!image) {
		return SOURCEPP_BUFFER_INVALID;
	}
	*width = image->width;
	*height = image->height;
	return convert::toBuffer(image->data);
}

SOURCEPP_API int vtfpp_ppl_set_image_from_file(vtfpp_ppl_handle_t handle, const char* imagePath, uint32_t lod, float quality) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(imagePath, false);

	return Convert::ppl(handle)->setImage(imagePath, lod, quality);
}

SOURCEPP_API int vtfpp_ppl_set_image_from_file_with_options(vtfpp_ppl_handle_t handle, const char* imagePath, uint32_t resizedWidth, uint32_t resizedHeight, uint32_t lod, vtfpp_image_conversion_resize_filter_e filter, float quality) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(imagePath, false);
	SOURCEPP_EARLY_RETURN_VAL(resizedWidth, false);
	SOURCEPP_EARLY_RETURN_VAL(resizedHeight, false);

	return Convert::ppl(handle)->setImage(imagePath, resizedWidth, resizedHeight, lod, static_cast<ImageConversion::ResizeFilter>(filter), quality);
}

SOURCEPP_API int vtfpp_ppl_set_image_from_mem(vtfpp_ppl_handle_t handle, const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint32_t width, uint32_t height, uint32_t lod, float quality) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(buffer, false);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, false);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, false);
	SOURCEPP_EARLY_RETURN_VAL(width, false);
	SOURCEPP_EARLY_RETURN_VAL(height, false);

	return Convert::ppl(handle)->setImage({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height, lod, quality);
}

SOURCEPP_API int vtfpp_ppl_set_image_from_mem_with_options(vtfpp_ppl_handle_t handle, const unsigned char* buffer, size_t bufferLen, vtfpp_image_format_e format, uint32_t width, uint32_t height, uint32_t resizedWidth, uint32_t resizedHeight, uint32_t lod, vtfpp_image_conversion_resize_filter_e filter, float quality) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(buffer, false);
	SOURCEPP_EARLY_RETURN_VAL(bufferLen, false);
	SOURCEPP_EARLY_RETURN_VAL(format != VTFPP_IMAGE_FORMAT_EMPTY, false);
	SOURCEPP_EARLY_RETURN_VAL(width, false);
	SOURCEPP_EARLY_RETURN_VAL(height, false);
	SOURCEPP_EARLY_RETURN_VAL(resizedWidth, false);
	SOURCEPP_EARLY_RETURN_VAL(resizedHeight, false);

	return Convert::ppl(handle)->setImage({reinterpret_cast<const std::byte*>(buffer), bufferLen}, static_cast<ImageFormat>(format), width, height, resizedWidth, resizedHeight, lod, static_cast<ImageConversion::ResizeFilter>(filter), quality);
}

SOURCEPP_API sourcepp_buffer_t vtfpp_ppl_save_image_to_file(vtfpp_ppl_handle_t handle, const char* imagePath, uint32_t lod, vtfpp_image_conversion_file_format_e fileFormat) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);
	SOURCEPP_EARLY_RETURN_VAL(imagePath, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(Convert::ppl(handle)->saveImageToFile(lod, static_cast<ImageConversion::FileFormat>(fileFormat)));
}

SOURCEPP_API sourcepp_buffer_t vtfpp_ppl_bake(vtfpp_ppl_handle_t handle) {
	SOURCEPP_EARLY_RETURN_VAL(handle, SOURCEPP_BUFFER_INVALID);

	return convert::toBuffer(Convert::ppl(handle)->bake());
}

SOURCEPP_API int vtfpp_ppl_bake_to_file(vtfpp_ppl_handle_t handle, const char* pplPath) {
	SOURCEPP_EARLY_RETURN_VAL(handle, false);
	SOURCEPP_EARLY_RETURN_VAL(pplPath, false);

	return Convert::ppl(handle)->bake(pplPath);
}

SOURCEPP_API void vtfpp_ppl_close(vtfpp_ppl_handle_t* handle) {
	SOURCEPP_EARLY_RETURN(handle);

	delete Convert::ppl(*handle);
	*handle = nullptr;
}

File: lang/csharp/src/vtfpp/PPL.cs
using System;

namespace sourcepp.vtfpp;

public sealed class PPL : sourcepp.ManagedNativeHandle
{
	internal PPL(nint handle) : base(handle, DLL.vtfpp_ppl_close)
	{
	}

	public PPL(uint modelChecksum, ImageFormat format = ImageFormat.RGB888, uint version = 0) : this(DLL.vtfpp_ppl_create_with_options(modelChecksum, format, version))
	{
	}

	public PPL(ReadOnlySpan<byte> buffer) : this(DLL.vtfpp_ppl_open_from_mem(buffer, (ulong) buffer.Length))
	{
	}

	public PPL(string pplPath) : this(DLL.vtfpp_ppl_open_from_file(pplPath))
	{
	}

	public static implicit operator bool(PPL ppl)
	{
		ppl.ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_is_valid(ppl.Handle));
	}

	public uint Version
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vtfpp_ppl_get_version(Handle);
		}
		set
		{
			ThrowIfDisposed();
			DLL.vtfpp_ppl_set_version(Handle, value);
		}
	}

	public uint ModelChecksum
	{
		get
		{
			ThrowIfDisposed();
			return DLL.vtfpp_ppl_get_model_checksum(Handle);
		}
		set
		{
			ThrowIfDisposed();
			DLL.vtfpp_ppl_set_model_checksum(Handle, value);
		}
	}

	public ImageFormat Format()
	{
		ThrowIfDisposed();
		return DLL.vtfpp_ppl_get_format(Handle);
	}

	public void Format(ImageFormat format, float quality = ImageConversion.DEFAULT_COMPRESSED_QUALITY)
	{
		ThrowIfDisposed();
		DLL.vtfpp_ppl_set_format(Handle, format, quality);
	}

	public bool HasImageForLod(uint lod)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_has_image_for_lod(Handle, lod));
	}

	public uint[] ImageLods
	{
		get
		{
			ThrowIfDisposed();
			return new sourcepp.Buffer(DLL.vtfpp_ppl_get_image_lods(Handle)).Read<uint>();
		}
	}

	public Span<byte> GetImageRaw(ref uint width, ref uint height, uint lod = 0)
	{
		ThrowIfDisposed();
		unsafe
		{
			ulong imageLen = 0;
			var imagePtr = DLL.vtfpp_ppl_get_image_raw(Handle, ref imageLen, ref width, ref height, lod);
			return new Span<byte>(imagePtr, (int) imageLen);
		}
	}

	public byte[] GetImageAs(ref uint width, ref uint height, ImageFormat format, uint lod = 0)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_ppl_get_image_as(Handle, ref width, ref height, format, lod)).Read<byte>();
	}
	
	// ReSharper disable once InconsistentNaming
	public byte[] GetImageAsRGB888(ref uint width, ref uint height, uint lod = 0)
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_ppl_get_image_as_rgb888(Handle, ref width, ref height, lod)).Read<byte>();
	}

	public bool SetImage(string imagePath, uint lod = 0, float quality = ImageConversion.DEFAULT_COMPRESSED_QUALITY)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_set_image_from_file(Handle, imagePath, lod, quality));
	}
	
	public bool SetImage(string imagePath, uint resizedWidth, uint resizedHeight, uint lod = 0, ImageConversion.ResizeFilter filter = ImageConversion.ResizeFilter.DEFAULT, float quality = ImageConversion.DEFAULT_COMPRESSED_QUALITY)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_set_image_from_file_with_options(Handle, imagePath, resizedWidth, resizedHeight, lod, filter, quality));
	}

	public bool SetImage(ReadOnlySpan<byte> buffer, ImageFormat format, uint width, uint height, uint lod = 0, float quality = ImageConversion.DEFAULT_COMPRESSED_QUALITY)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_set_image_from_mem(Handle, buffer, (ulong) buffer.Length, format, width, height, lod, quality));
	}
	
	public bool SetImage(ReadOnlySpan<byte> buffer, ImageFormat format, uint width, uint height, uint resizedWidth, uint resizedHeight, uint lod = 0, ImageConversion.ResizeFilter filter = ImageConversion.ResizeFilter.DEFAULT, float quality = ImageConversion.DEFAULT_COMPRESSED_QUALITY)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_set_image_from_mem_with_options(Handle, buffer, (ulong) buffer.Length, format, width, height, resizedWidth, resizedHeight, lod, filter, quality));
	}

	public bool RemoveImage(uint lod)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_remove_image(Handle, lod));
	}

	public bool SaveImageToFile(string imagePath, uint lod, ImageConversion.FileFormat fileFormat = ImageConversion.FileFormat.DEFAULT)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_save_image_to_file(Handle, imagePath, lod, fileFormat));
	}

	public byte[] Bake()
	{
		ThrowIfDisposed();
		return new sourcepp.Buffer(DLL.vtfpp_ppl_bake(Handle)).Read<byte>();
	}

	public bool Bake(string pplPath)
	{
		ThrowIfDisposed();
		return Convert.ToBoolean(DLL.vtfpp_ppl_bake_to_file(Handle, pplPath));
	}
}
